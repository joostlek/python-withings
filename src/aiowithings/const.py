"""Asynchronous Python client for Withings."""
from __future__ import annotations

import logging

LOGGER = logging.getLogger(__package__)

STATUS_SUCCESS: set[int] = {0}

STATUS_AUTH_FAILED: set[int] = {100, 101, 102, 200, 401}

STATUS_INVALID_PARAMS: set[int] = {
    201,
    202,
    203,
    204,
    205,
    206,
    207,
    208,
    209,
    210,
    211,
    212,
    213,
    216,
    217,
    218,
    220,
    221,
    223,
    225,
    227,
    228,
    229,
    230,
    234,
    235,
    236,
    238,
    240,
    241,
    242,
    243,
    244,
    245,
    246,
    247,
    248,
    249,
    250,
    251,
    252,
    254,
    260,
    261,
    262,
    263,
    264,
    265,
    266,
    267,
    271,
    272,
    275,
    276,
    283,
    284,
    285,
    286,
    287,
    288,
    290,
    293,
    294,
    295,
    297,
    300,
    301,
    302,
    303,
    304,
    321,
    323,
    324,
    325,
    326,
    327,
    328,
    329,
    330,
    331,
    332,
    333,
    334,
    335,
    336,
    337,
    338,
    339,
    340,
    341,
    342,
    343,
    344,
    345,
    346,
    347,
    348,
    349,
    350,
    351,
    352,
    353,
    380,
    381,
    382,
    400,
    501,
    502,
    503,
    504,
    505,
    506,
    509,
    510,
    511,
    523,
    532,
    3017,
    3018,
    3019,
}

STATUS_UNAUTHORIZED: set[int] = {214, 277, 2553, 2554, 2555}

STATUS_ERROR_OCCURRED: set[int] = {
    215,
    219,
    222,
    224,
    226,
    231,
    233,
    237,
    253,
    255,
    256,
    257,
    258,
    259,
    268,
    269,
    270,
    273,
    274,
    278,
    279,
    280,
    281,
    282,
    289,
    291,
    292,
    296,
    298,
    305,
    306,
    308,
    309,
    310,
    311,
    312,
    313,
    314,
    315,
    316,
    317,
    318,
    319,
    320,
    322,
    370,
    371,
    372,
    373,
    374,
    375,
    383,
    391,
    402,
    516,
    517,
    518,
    519,
    520,
    521,
    525,
    526,
    527,
    528,
    529,
    530,
    531,
    533,
    602,
    700,
    1051,
    1052,
    1053,
    1054,
    2551,
    2552,
    2556,
    2557,
    2558,
    2559,
    3000,
    3001,
    3002,
    3003,
    3004,
    3005,
    3006,
    3007,
    3008,
    3009,
    3010,
    3011,
    3012,
    3013,
    3014,
    3015,
    3016,
    3020,
    3021,
    3022,
    3023,
    3024,
    5000,
    5001,
    5005,
    5006,
    6000,
    6010,
    6011,
    9000,
    10000,
}

STATUS_TIMEOUT: set[int] = {522}
STATUS_BAD_STATE: set[int] = {524}
STATUS_TOO_MANY_REQUESTS: set[int] = {601}
