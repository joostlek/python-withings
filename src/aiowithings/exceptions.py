"""Asynchronous Python client for Withings."""


class WithingsError(Exception):
    """Generic exception."""


class WithingsConnectionError(WithingsError):
    """Withings connection exception."""


class WithingsAuthenticationFailedError(WithingsError):
    """Withings authentication failed exception."""


class WithingsInvalidParamsError(WithingsError):
    """Withings invalid params exception."""


class WithingsUnauthorizedError(WithingsError):
    """Withings unauthorized exception."""


class WithingsErrorOccurredError(WithingsError):
    """Withings error occurred exception."""


class WithingsBadStateError(WithingsError):
    """Withings bad state exception."""


class WithingsTooManyRequestsError(WithingsError):
    """Withings too many requests exception."""


class WithingsUnknownStatusError(WithingsError):
    """Withings unknown status exception."""
