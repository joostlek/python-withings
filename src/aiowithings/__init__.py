"""Asynchronous Python client for Withings."""
from .const import AUTHORIZATION_URL, TOKEN_URL
from .exceptions import (
    WithingsAuthenticationFailedError,
    WithingsBadStateError,
    WithingsConnectionError,
    WithingsError,
    WithingsErrorOccurredError,
    WithingsInvalidParamsError,
    WithingsTooManyRequestsError,
    WithingsUnauthorizedError,
    WithingsUnknownStatusError,
)
from .helpers import aggregate_measurements
from .models import (
    Activity,
    ActivityDataFields,
    ActivityDataOrigin,
    AuthScope,
    Device,
    DeviceBattery,
    DeviceModel,
    DeviceType,
    Goals,
    Measurement,
    MeasurementAttribution,
    MeasurementGroup,
    MeasurementGroupCategory,
    MeasurementType,
    NotificationCategory,
    Services,
    SleepDataFields,
    SleepSeries,
    SleepSeriesTimeData,
    SleepState,
    SleepSummary,
    SleepSummaryDataFields,
    WebhookCall,
    Workout,
    WorkoutCategory,
    WorkoutDataFields,
    get_measurement_type_from_notification_category,
)
from .withings import WithingsClient

__all__ = [
    "AUTHORIZATION_URL",
    "TOKEN_URL",
    "AuthScope",
    "ActivityDataFields",
    "ActivityDataOrigin",
    "Activity",
    "DeviceBattery",
    "Device",
    "DeviceType",
    "DeviceModel",
    "WithingsError",
    "WithingsConnectionError",
    "WithingsAuthenticationFailedError",
    "WithingsInvalidParamsError",
    "WithingsUnauthorizedError",
    "WithingsErrorOccurredError",
    "WithingsBadStateError",
    "WithingsTooManyRequestsError",
    "WithingsUnknownStatusError",
    "WithingsClient",
    "Goals",
    "MeasurementAttribution",
    "MeasurementGroupCategory",
    "MeasurementGroup",
    "MeasurementType",
    "Measurement",
    "get_measurement_type_from_notification_category",
    "aggregate_measurements",
    "NotificationCategory",
    "Services",
    "WebhookCall",
    "SleepSummary",
    "SleepSummaryDataFields",
    "SleepState",
    "SleepSeriesTimeData",
    "SleepSeries",
    "SleepDataFields",
    "WorkoutDataFields",
    "Workout",
    "WorkoutCategory",
]
