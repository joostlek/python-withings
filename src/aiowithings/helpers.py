"""Asynchronous Python client for Withings."""
from __future__ import annotations

from typing import TYPE_CHECKING

from aiowithings.models import MeasurementAttribution, MeasurementType, SleepSummary

if TYPE_CHECKING:
    from aiowithings import MeasurementGroup


def aggregate_measurements(
    measurements: list[MeasurementGroup],
) -> dict[MeasurementType, float]:
    """Aggregate the measurements to return a list of the latest measurements."""
    result: dict[MeasurementType, float] = {}

    groups = sorted(measurements, key=lambda group: group.taken_at)

    for measurement in groups:
        if measurement.attribution not in (
            MeasurementAttribution.UNKNOWN,
            MeasurementAttribution.DEVICE_ENTRY_FOR_USER_AMBIGUOUS,
        ):
            for data_point in measurement.measurements:
                result[data_point.measurement_type] = data_point.value

    return result


def aggregate_sleep_summary(
    sleep_summaries: list[SleepSummary],
) -> SleepSummary | None:
    """Aggregate the sleep summaries to return a combined report."""
    if len(sleep_summaries) == 0:
        return None
    if len(sleep_summaries) == 1:
        return sleep_summaries[0]
    sorted_sleep_summaries = sorted(
        sleep_summaries,
        key=lambda sleep_summary: sleep_summary.start_date,
    )

    def average(data: list[int]) -> int:
        return round(sum(data) / len(data))

    return SleepSummary(
        start_date=sorted_sleep_summaries[0].start_date,
        end_date=sorted_sleep_summaries[1].end_date,
        date=sorted_sleep_summaries[0].date,
        hashed_device_id=sorted_sleep_summaries[0].hashed_device_id,
        breathing_disturbances_intensity=average(
            [
                summary.breathing_disturbances_intensity
                for summary in sleep_summaries
                if summary.breathing_disturbances_intensity is not None
            ],
        ),
        deep_sleep_duration=sum(
            summary.deep_sleep_duration
            for summary in sleep_summaries
            if summary.deep_sleep_duration is not None
        ),
        sleep_latency=average(
            [
                summary.sleep_latency
                for summary in sleep_summaries
                if summary.sleep_latency is not None
            ],
        ),
        wake_up_latency=average(
            [
                summary.wake_up_latency
                for summary in sleep_summaries
                if summary.wake_up_latency is not None
            ],
        ),
        average_heart_rate=average(
            [
                summary.average_heart_rate
                for summary in sleep_summaries
                if summary.average_heart_rate is not None
            ],
        ),
        max_heart_rate=average(
            [
                summary.max_heart_rate
                for summary in sleep_summaries
                if summary.max_heart_rate is not None
            ],
        ),
        min_heart_rate=average(
            [
                summary.min_heart_rate
                for summary in sleep_summaries
                if summary.min_heart_rate is not None
            ],
        ),
        light_sleep_duration=sum(
            summary.light_sleep_duration
            for summary in sleep_summaries
            if summary.light_sleep_duration is not None
        ),
        rem_sleep_duration=sum(
            summary.rem_sleep_duration
            for summary in sleep_summaries
            if summary.rem_sleep_duration is not None
        ),
        average_respiration_rate=average(
            [
                summary.average_respiration_rate
                for summary in sleep_summaries
                if summary.average_respiration_rate is not None
            ],
        ),
        max_respiration_rate=average(
            [
                summary.max_respiration_rate
                for summary in sleep_summaries
                if summary.max_respiration_rate is not None
            ],
        ),
        min_respiration_rate=average(
            [
                summary.min_respiration_rate
                for summary in sleep_summaries
                if summary.min_respiration_rate is not None
            ],
        ),
        sleep_score=max(
            (
                summary.sleep_score
                for summary in sleep_summaries
                if summary.sleep_score is not None
            ),
            default=0,
        ),
        snoring=average(
            [
                summary.snoring
                for summary in sleep_summaries
                if summary.snoring is not None
            ],
        ),
        snoring_count=sum(
            summary.snoring_count
            for summary in sleep_summaries
            if summary.snoring_count is not None
        ),
        wake_up_count=sum(
            summary.wake_up_count
            for summary in sleep_summaries
            if summary.wake_up_count is not None
        ),
        total_time_awake=average(
            [
                summary.total_time_awake
                for summary in sleep_summaries
                if summary.total_time_awake is not None
            ],
        ),
        active_movement_duration=None,
        apnea_hypopnea_index=None,
        average_movement_score=None,
        external_time_asleep=None,
        out_of_bed_count=None,
        rem_sleep_phase_count=None,
        sleep_efficiency=None,
        time_awake_during_sleep=None,
        total_sleep_time=None,
        total_time_in_bed=None,
        withings_index=None,
    )
