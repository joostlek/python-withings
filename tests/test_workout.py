"""Asynchronous Python client for Withings."""
from datetime import datetime, timezone

import aiohttp
from aresponses import ResponsesMockServer

from aiowithings import WithingsClient
from aiowithings.models import WorkoutDataFields
from syrupy import SnapshotAssertion

from . import load_fixture

WITHINGS_URL = "wbsapi.withings.net"


async def test_get_workouts_since(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving workouts."""
    aresponses.add(
        WITHINGS_URL,
        "/v2/measure",
        "POST",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("workouts.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        withings = WithingsClient(session=session)
        withings.authenticate("test")
        response = await withings.get_workouts_since(
            datetime.now(tz=timezone.utc),
            workout_data_fields=[WorkoutDataFields.CALORIES],
        )
        assert response == snapshot
        await withings.close()


async def test_get_workouts_period(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving workouts."""
    aresponses.add(
        WITHINGS_URL,
        "/v2/measure",
        "POST",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("workouts.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        withings = WithingsClient(session=session)
        withings.authenticate("test")
        response = await withings.get_workouts_in_period(
            start_date=datetime.now(tz=timezone.utc),
            end_date=datetime.now(tz=timezone.utc),
        )
        assert response == snapshot
        await withings.close()
