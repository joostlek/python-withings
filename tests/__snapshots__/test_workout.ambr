# serializer version: 1
# name: test_get_workouts_period
  list([
    dict({
      'active_calories_burnt': 47,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': 80,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 8, 29),
      'distance': 680,
      'duration_heart_rate_intense_zone': 300,
      'duration_heart_rate_light_zone': 100,
      'duration_heart_rate_maximal_zone': 400,
      'duration_heart_rate_moderate_zone': 200,
      'end_date': datetime.datetime(2023, 8, 29, 19, 15, 13, tzinfo=datetime.timezone.utc),
      'floors_climbed': 10,
      'intensity': 30,
      'max_heart_rate': 80,
      'min_heart_rate': 70,
      'pause_duration': 80,
      'spo2_average': 15,
      'start_date': datetime.datetime(2023, 8, 29, 19, 6, 51, tzinfo=datetime.timezone.utc),
      'steps': 779,
      'workout_id': 3661300277,
    }),
    dict({
      'active_calories_burnt': None,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 8, 31),
      'distance': None,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 8, 31, 8, 18, 44, tzinfo=datetime.timezone.utc),
      'floors_climbed': None,
      'intensity': None,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 8, 31, 8, 8, 27, tzinfo=datetime.timezone.utc),
      'steps': None,
      'workout_id': 3661300290,
    }),
    dict({
      'active_calories_burnt': 82,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 8, 4),
      'distance': 1294,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 8, 4, 16, 15, 19, tzinfo=datetime.timezone.utc),
      'floors_climbed': 18,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 8, 4, 16, 0, 39, tzinfo=datetime.timezone.utc),
      'steps': 1450,
      'workout_id': 3661300269,
    }),
    dict({
      'active_calories_burnt': 97,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 9, 23),
      'distance': 1405,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 9, 22, 23, 51, 1, tzinfo=datetime.timezone.utc),
      'floors_climbed': 19,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 9, 22, 23, 33, 55, tzinfo=datetime.timezone.utc),
      'steps': 1650,
      'workout_id': 3743596080,
    }),
    dict({
      'active_calories_burnt': 62,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 9, 14),
      'distance': 917,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 9, 14, 18, 31, 46, tzinfo=datetime.timezone.utc),
      'floors_climbed': 15,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 9, 14, 18, 20, 49, tzinfo=datetime.timezone.utc),
      'steps': 1076,
      'workout_id': 3743596073,
    }),
    dict({
      'active_calories_burnt': 13,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 9, 23),
      'distance': 185,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 9, 22, 23, 58, 13, tzinfo=datetime.timezone.utc),
      'floors_climbed': 4,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 9, 22, 23, 55, 53, tzinfo=datetime.timezone.utc),
      'steps': 216,
      'workout_id': 3743596085,
    }),
    dict({
      'active_calories_burnt': 187,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 9, 14),
      'distance': 2908,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 9, 14, 18, 15, 27, tzinfo=datetime.timezone.utc),
      'floors_climbed': 49,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 9, 14, 17, 42, 31, tzinfo=datetime.timezone.utc),
      'steps': 3339,
      'workout_id': 3743596072,
    }),
    dict({
      'active_calories_burnt': 18,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 10, 9),
      'distance': 261,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 10, 9, 7, 16, 7, tzinfo=datetime.timezone.utc),
      'floors_climbed': 4,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 10, 9, 7, 12, 49, tzinfo=datetime.timezone.utc),
      'steps': 291,
      'workout_id': 3752609171,
    }),
    dict({
      'active_calories_burnt': 24,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 10, 9),
      'distance': 232,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 10, 9, 9, 43, 58, tzinfo=datetime.timezone.utc),
      'floors_climbed': 4,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 10, 9, 9, 39, 43, tzinfo=datetime.timezone.utc),
      'steps': 267,
      'workout_id': 3752609178,
    }),
    dict({
      'active_calories_burnt': 21,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 10, 9),
      'distance': 359,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 10, 9, 9, 17, 12, tzinfo=datetime.timezone.utc),
      'floors_climbed': 4,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 10, 9, 9, 13, 23, tzinfo=datetime.timezone.utc),
      'steps': 403,
      'workout_id': 3752609174,
    }),
    dict({
      'active_calories_burnt': 21,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 10, 9),
      'distance': 359,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 10, 9, 9, 17, 12, tzinfo=datetime.timezone.utc),
      'floors_climbed': 4,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 10, 9, 9, 13, 23, tzinfo=datetime.timezone.utc),
      'steps': 403,
      'workout_id': 3752609174,
    }),
  ])
# ---
# name: test_get_workouts_since
  list([
    dict({
      'active_calories_burnt': 47,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': 80,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 8, 29),
      'distance': 680,
      'duration_heart_rate_intense_zone': 300,
      'duration_heart_rate_light_zone': 100,
      'duration_heart_rate_maximal_zone': 400,
      'duration_heart_rate_moderate_zone': 200,
      'end_date': datetime.datetime(2023, 8, 29, 19, 15, 13, tzinfo=datetime.timezone.utc),
      'floors_climbed': 10,
      'intensity': 30,
      'max_heart_rate': 80,
      'min_heart_rate': 70,
      'pause_duration': 80,
      'spo2_average': 15,
      'start_date': datetime.datetime(2023, 8, 29, 19, 6, 51, tzinfo=datetime.timezone.utc),
      'steps': 779,
      'workout_id': 3661300277,
    }),
    dict({
      'active_calories_burnt': None,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 8, 31),
      'distance': None,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 8, 31, 8, 18, 44, tzinfo=datetime.timezone.utc),
      'floors_climbed': None,
      'intensity': None,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 8, 31, 8, 8, 27, tzinfo=datetime.timezone.utc),
      'steps': None,
      'workout_id': 3661300290,
    }),
    dict({
      'active_calories_burnt': 82,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 8, 4),
      'distance': 1294,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 8, 4, 16, 15, 19, tzinfo=datetime.timezone.utc),
      'floors_climbed': 18,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 8, 4, 16, 0, 39, tzinfo=datetime.timezone.utc),
      'steps': 1450,
      'workout_id': 3661300269,
    }),
    dict({
      'active_calories_burnt': 97,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 9, 23),
      'distance': 1405,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 9, 22, 23, 51, 1, tzinfo=datetime.timezone.utc),
      'floors_climbed': 19,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 9, 22, 23, 33, 55, tzinfo=datetime.timezone.utc),
      'steps': 1650,
      'workout_id': 3743596080,
    }),
    dict({
      'active_calories_burnt': 62,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 9, 14),
      'distance': 917,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 9, 14, 18, 31, 46, tzinfo=datetime.timezone.utc),
      'floors_climbed': 15,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 9, 14, 18, 20, 49, tzinfo=datetime.timezone.utc),
      'steps': 1076,
      'workout_id': 3743596073,
    }),
    dict({
      'active_calories_burnt': 13,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 9, 23),
      'distance': 185,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 9, 22, 23, 58, 13, tzinfo=datetime.timezone.utc),
      'floors_climbed': 4,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 9, 22, 23, 55, 53, tzinfo=datetime.timezone.utc),
      'steps': 216,
      'workout_id': 3743596085,
    }),
    dict({
      'active_calories_burnt': 187,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 9, 14),
      'distance': 2908,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 9, 14, 18, 15, 27, tzinfo=datetime.timezone.utc),
      'floors_climbed': 49,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 9, 14, 17, 42, 31, tzinfo=datetime.timezone.utc),
      'steps': 3339,
      'workout_id': 3743596072,
    }),
    dict({
      'active_calories_burnt': 18,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 10, 9),
      'distance': 261,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 10, 9, 7, 16, 7, tzinfo=datetime.timezone.utc),
      'floors_climbed': 4,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 10, 9, 7, 12, 49, tzinfo=datetime.timezone.utc),
      'steps': 291,
      'workout_id': 3752609171,
    }),
    dict({
      'active_calories_burnt': 24,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 10, 9),
      'distance': 232,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 10, 9, 9, 43, 58, tzinfo=datetime.timezone.utc),
      'floors_climbed': 4,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 10, 9, 9, 39, 43, tzinfo=datetime.timezone.utc),
      'steps': 267,
      'workout_id': 3752609178,
    }),
    dict({
      'active_calories_burnt': 21,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 10, 9),
      'distance': 359,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 10, 9, 9, 17, 12, tzinfo=datetime.timezone.utc),
      'floors_climbed': 4,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 10, 9, 9, 13, 23, tzinfo=datetime.timezone.utc),
      'steps': 403,
      'workout_id': 3752609174,
    }),
    dict({
      'active_calories_burnt': 21,
      'attribution': <MeasurementAttribution.DEVICE_ENTRY_FOR_USER: 0>,
      'average_heart_rate': None,
      'category': <WorkoutCategory.WALK: 1>,
      'date': datetime.date(2023, 10, 9),
      'distance': 359,
      'duration_heart_rate_intense_zone': None,
      'duration_heart_rate_light_zone': None,
      'duration_heart_rate_maximal_zone': None,
      'duration_heart_rate_moderate_zone': None,
      'end_date': datetime.datetime(2023, 10, 9, 9, 17, 12, tzinfo=datetime.timezone.utc),
      'floors_climbed': 4,
      'intensity': 30,
      'max_heart_rate': None,
      'min_heart_rate': None,
      'pause_duration': None,
      'spo2_average': None,
      'start_date': datetime.datetime(2023, 10, 9, 9, 13, 23, tzinfo=datetime.timezone.utc),
      'steps': 403,
      'workout_id': 3752609174,
    }),
  ])
# ---
